docs/pages/pmd/userdocs/tools/java-api.md

```diff
@@ -80,71 +80,77 @@ public class PmdExample {
+    configuration.prependAuxClasspath("/home/workspace/target/classes:/home/.m2/repository/my/dependency.jar");
     ```
     
-4.  PMD operates on a list of `DataSource`. You can assemble a own list of `FileDataSource`, e.g.
+4.  Then we configure the reporting. Configuring the report file is optional. If not specified, the report
+    will be written to `stdout`.
     
     ```java
-    List<DataSource> files = Arrays.asList(new FileDataSource(new File("/path/to/src/MyClass.java")));
+    configuration.setReportFormat("xml");
+    configuration.setReportFile("/home/workspace/pmd-report.xml");
     ```
     
-5.  For reporting, you can use a built-in renderer, e.g. `XMLRenderer` or a custom renderer implementing
-    `Renderer`. Note, that you must manually initialize
-    the renderer by setting a suitable `Writer` and calling `start()`. After the PMD run, you need to call
-    `end()` and `flush()`. Then your writer should have received all output.
+5.  Now an optional step: If you want to use additional renderers as in the example, set them up before
+    calling PMD. You can use a built-in renderer, e.g. `XMLRenderer` or a custom renderer implementing
+    `Renderer`. Note, that you must manually initialize the renderer by setting a suitable `Writer`:
     
     ```java
-    StringWriter rendererOutput = new StringWriter();
-    Renderer xmlRenderer = new XMLRenderer("UTF-8");
-    xmlRenderer.setWriter(rendererOutput);
-    xmlRenderer.start();
+    Writer rendererOutput = new StringWriter();
+    Renderer renderer = createRenderer(rendererOutput);
+    
+    // ...
+    private static Renderer createRenderer(Writer writer) {
+        XMLRenderer xml = new XMLRenderer("UTF-8");
+        xml.setWriter(writer);
+        return xml;
+    }
     ```
     
-6.  Now, all the preparations are done, and PMD can be executed. This is done by calling
-    `PMD.processFiles(...)`. This method call takes the configuration, the rulesets, the files
-    to process, and the list of renderers. Provide an empty list, if you don't want to use
-    any renderer. Note: The auxclasspath needs to be closed explicitly. Otherwise the class or jar files may
-    remain open and file resources are leaked.
+6.  Finally we can start the PMD analysis. There is the possibility to fine-tune the configuration
+    by adding additional files to analyze or adding additional rulesets or renderers:
     
     ```java
-    try {
-        PMD.processFiles(configuration, ruleSets, files, Collections.singletonList(renderer));
-    } finally {
-        ClassLoader auxiliaryClassLoader = configuration.getClassLoader();
-        if (auxiliaryClassLoader instanceof ClasspathClassLoader) {
-            ((ClasspathClassLoader) auxiliaryClassLoader).close();
-        }
+    try (PmdAnalysis pmd = PmdAnalysis.create(configuration)) {
+        // optional: add more rulesets
+        pmd.addRuleSet(RuleSetLoader.fromPmdConfig(configuration).loadFromResource("custom-ruleset.xml"));
```

Comment:
I wonder if PmdAnalysis should expose a  `newRuleSetLoader` method. This looks like a roundabout way

Reply:
I think PmdAnalysis should gradually overtake the responsibilities of PMDConfiguration. Currently PMDConfiguration can be regarded as an "initial configuration" for PmdAnalysis, a user shouldn't manipulate the configuration if they already have a PmdAnalysis.